Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
     return true;
 }
 
+// ------------------------------ FastStringifier --------------------------------
+
+// FastStringifier does a no-side-effects stringify of the most common types of
+// objects and arrays. It bails out if the serialization is any longer than a
+// fixed buffer and handles only the simplest cases, including only 8-bit character
+// strings. Instead of explicit checks to prevent excessive recursion and cycles,
+// it counts on hitting the buffer size limit to catch those things. If it fails,
+// since there is no side effect, the real Stringifier can be used and the only
+// cost of the fast stringifying attempt is the time/speed wasted.
+
+#if !FAST_STRINGIFY_LOG_USAGE
+
+static inline void addStringifyOutcome(ASCIILiteral)
+{
+}
+
+static inline void addStringifyBufferSize(unsigned)
+{
+}
+
+#else
+
+static void addStringifyOutcome(String&& outcome)
+{
+    static NeverDestroyed<HashCountedSet<String>> set;
+    static std::atomic<unsigned> count;
+    set->add(outcome);
+    if (!(++count % 100)) {
+        Vector<KeyValuePair<String, unsigned>> vector;
+        for (auto& pair : set.get())
+            vector.append(pair);
+        std::sort(vector.begin(), vector.end(), [](auto& a, auto &b) {
+            return a.value != b.value ? a.value > b.value : codePointCompareLessThan(a.key, b.key);
+        });
+        dataLogLn("fastStringify outcomes");
+        for (auto& pair : vector) {
+            dataLogF("%5u", pair.value);
+            dataLogLn(": ", pair.key);
+        }
+    }
+}
+
+static void addStringifyBufferSize(unsigned size)
+{
+    static std::atomic<unsigned> maxSizeSeen;
+    if (size <= maxSizeSeen)
+        return;
+    maxSizeSeen = size;
+    dataLogLn("max fastStringify buffer size used: ", size);
+}
+
+#endif
+
+class FastStringifier {
+public:
+    explicit FastStringifier(JSGlobalObject&);
+    void append(JSValue);
+    String result() const;
+
+private:
+    void recordBufferFull();
+    void recordFailure(ASCIILiteral reason);
+    void recordFailure(String&& reason);
+    void recordFastPropertyEnumerationFailure(JSObject&);
+    bool haveFailure() const;
+    unsigned remainingCapacity() const;
+    bool mayHaveToJSON(JSObject&) const;
+
+    JSGlobalObject& m_globalObject;
+    VM& m_vm;
+    unsigned m_length { 0 };
+    bool m_checkedObjectPrototype { false };
+    bool m_checkedArrayPrototype { false };
+
+    // Note that this buffer needs to be kept reasonably small; it acts as a recursion limit and a cycle detector as well.
+    LChar m_buffer[6000];
+};
+
+inline FastStringifier::FastStringifier(JSGlobalObject& globalObject)
+    : m_globalObject(globalObject)
+    , m_vm(globalObject.vm())
+{
+}
+
+inline bool FastStringifier::haveFailure() const
+{
+    return m_length > std::size(m_buffer);
+}
+
+inline String FastStringifier::result() const
+{
+    if (haveFailure())
+        return { };
+    addStringifyBufferSize(m_length);
```

Comment:
nit: calling this function `addStringBufferSize` seems strange because it isn't really adding anything together (which is what it suggested to me when I first read it).  What do you think of calling it `recordStringBufferSize`?  Ditto for `addStringifyOutcome` => `recordStringifyOutcome`.  Calling them `record` is also consistent

Reply:
Yes, good suggestion. I will rename them.

Reply:
Fixed locally.